### PR TITLE
Add GatheringICECandidate Finish Duration to Meeting Event

### PR DESCRIPTION
**Pull request by RidipDe**
**Issue #:**

**Description of changes:**
Add GatheringICECandidate Finish Duration to Meeting Event and to Demo App

**Testing**

1. Have you successfully run `npm run build:release` locally?
Yes

2. How did you test these changes?
Manually

3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it?
Yes, meeting app

4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
Yes

5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by simmkyu — CHANGES_REQUESTED** [review]
- Please add a description to the guide.
https://github.com/aws/amazon-chime-sdk-js/blob/master/guides/06_Meeting_Events.md#meeting-attributes
- Suggest renaming it to `iceGatheringDurationMs`
- Ignore this metric library in the [serverless .gitignore file](https://github.com/aws/amazon-chime-sdk-js/blob/master/demos/serverless/.gitignore). Otherwise, these files show up as untracked in a local Git repo.
  ```
  src/aws-embedded-metrics
  ```

File: src/audiovideocontroller/AudioVideoControllerState.ts [hunk 1]
```diff
@@ -120,6 +120,8 @@ export default class AudioVideoControllerState {
 
   signalingOpenDurationMs: number | null = null;
 
+  gatheringICECandidateFinishDurationMs: number | null = null;
```
[the remark this answers is not in the document]
Comment (RidipDe) [reply src/audiovideocontroller/AudioVideoControllerState.ts]: Changed gatheringICECandidateFinishDurationMs -> iceGatheringDurationMs

File: demos/serverless/src/handlers.js [hunk 2]
```diff
@@ -277,7 +277,7 @@ function response(statusCode, contentType, body) {
 
 function addSignalMetricsToCloudWatch(logMsg, meetingId, attendeeId) {
   const logMsgJson = JSON.parse(logMsg);
-  const metricList = ['signalingOpenDurationMs'];
+  const metricList = ['signalingOpenDurationMs', 'gatheringICECandidateFinishDurationMs'];
```
[the remark this answers is not in the document]
Comment (RidipDe) [reply demos/serverless/src/handlers.js]: CW Metrics treats null as 0. So no changes required.

File: guides/06_Meeting_Events.md [hunk 3]
```diff
@@ -127,6 +127,7 @@ The following table describes attributes for a meeting.
 |`poorConnectionCount`|The number of times the significant packet loss occurred during the meeting. Per count, you receive `AudioVideoObserver.connectionDidBecomePoor` or `AudioVideoObserver.connectionDidSuggestStopVideo`.<br><br>Unit: Count|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`
 |`retryCount`|The number of connection retries performed during the meeting.<br><br>Unit: Count|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`
 |`signalingOpenDurationMs`|The time taken for opening a WebSocket connection.<br><br>Unit: Milliseconds|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`
+|`iceGatheringDurationMs`|The time taken for connetion's ICE gathering state to complete.<br><br>Unit: Milliseconds|`meetingStartSucceeded`, `meetingStartFailed`, `meetingEnded`, `meetingFailed`
```
Comment (simmkyu) [inline guides/06_Meeting_Events.md:130]: Typo: `connetion` => `connection`

File: src/task/FinishGatheringICECandidatesTask.ts [hunk 4]
```diff
@@ -62,98 +62,103 @@ export default class FinishGatheringICECandidatesTask extends BaseTask {
...
-      this.startTimestampMs = Date.now();
-    });
+        // SDK does not catch candidate itself and send to sever. Rather, WebRTC handles candidate events and writes candidate to SDP.
+        this.context.peer.addEventListener('icecandidate', this.context.iceCandidateHandler);
+        this.startTimestampMs = Date.now();
+      });
+    } catch (error) {
+      throw error;
+    } finally {
+      this.context.iceGatheringDurationMs = Math.round(Date.now() - this.startTimestampMs);
```
Comment (simmkyu) [inline src/task/FinishGatheringICECandidatesTask.ts:161]: If the task returns at line 71 or 92 or throws anywhere before `this.startTimestampMs` is set, the `iceGatheringDurationMs` value might be `NaN`. You can solve in this way:

1. Use the try/catch statement to wrap only the promise.
2. To be defensive, also check if the `startTimestampMs` exists.

```
try {
  await new Promise<void>((resolve, reject) => {
     ...
  });
} catch (error) {
  throw error;
} finally {
  if (this.startTimestampMs) {
     this.context.iceGatheringDurationMs  = Math.round(Date.now() - this.startTimestampMs);
  }
}
```

**Review by simmkyu — CHANGES_REQUESTED** [review]

**Review by simmkyu — APPROVED** [review]

**Review by simmkyu — APPROVED** [review]

**Pull request merged**